uspace/lib/c/generic/async/server.c

```diff
@@ -662,33 +686,144 @@ static bool route_call(cap_call_handle_t chandle, ipc_call_t *call)
+
+		futex_lock(&notification_futex);
+
+		/*
+		 * The semaphore ensures that if we get this far,
+		 * the queue must be non-empty.
+		 */
+		assert(!list_empty(&notification_queue));
+
+		notification_t *notification = list_get_instance(
+		    list_first(&notification_queue), notification_t, qlink);
+		list_remove(&notification->qlink);
+
+		async_notification_handler_t handler = notification->handler;
+		void *data = notification->data;
+		ipc_call_t calldata = notification->calldata;
+		long count = notification->count;
+
+		notification->count = 0;
+
+		futex_unlock(&notification_futex);
+
+		// FIXME: Pass count to the handler. It might be important.
+		(void) count;
+
+		if (handler)
+			handler(&calldata, data);
+	}
+
+	/* Not reached. */
+	return EOK;
+}
+
+/**
+ * Creates a new dedicated fibril for handling notifications.
+ * By default, there is one such fibril. This function can be used to
+ * create more in order to increase the number of notification that can
+ * be processed concurrently.
+ *
+ * Currently, there is no way to destroy those fibrils after they are created.
+ */
+errno_t async_spawn_notification_handler()
+{
+	fid_t f = fibril_create(notification_fibril_func, NULL);
+	if (f == 0)
+		return ENOMEM;
+
+	fibril_add_ready(f);
+	return EOK;
+}
+
+/** Queue notification.
  *
  * @param call   Data of the incoming call.
  *
  */
-static void process_notification(ipc_call_t *call)
+static void queue_notification(ipc_call_t *call)
 {
-	async_notification_handler_t handler = NULL;
-	void *data = NULL;
-
 	assert(call);
 
-	futex_down(&async_futex);
+	futex_lock(&notification_futex);
 
 	ht_link_t *link = hash_table_find(&notification_hash_table,
 	    &IPC_GET_IMETHOD(*call));
-	if (link) {
-		notification_t *notification =
-		    hash_table_get_inst(link, notification_t, link);
-		handler = notification->handler;
-		data = notification->data;
+	if (!link) {
+		/* Invalid notification. */
+		// FIXME: Make sure this can't happen and turn it into assert.
+		futex_unlock(&notification_futex);
+		return;
 	}
 
-	futex_up(&async_futex);
+	notification_t *notification =
+	    hash_table_get_inst(link, notification_t, htlink);
+
+	notification->count++;
+	notification->calldata = *call;
+
+	if (link_in_use(&notification->qlink)) {
+		/* Notification already queued. */
+		futex_unlock(&notification_futex);
+		return;
+	}
+
+	list_append(&notification->qlink, &notification_queue);
+	futex_unlock(&notification_futex);
+
+	fibril_semaphore_up(&notification_semaphore);
+}
+
+static notification_t *notification_create(async_notification_handler_t handler, void *data)
+{
+	notification_t *notification = calloc(1, sizeof(notification_t));
+	if (!notification)
+		return NULL;
+
+	notification->handler = handler;
+	notification->data = data;
+
+	fid_t fib = 0;
+
+	futex_lock(&notification_futex);
+
+	if (notification_avail == 0) {
+		/* Attempt to create the first handler fibril. */
+		futex_unlock(&notification_futex);
```

Comment:
Why do we unlock the notification futex here?

Reply:
I'm not entirely sure. Probably just my deep seated discomfort with running expensive operations in a critical section. But it shouldn't make any difference in this case. I'll remove the unlock.